clippy_lints/src/path_buf_push_overwrite.rs

```diff
@@ -0,0 +1,66 @@
+use crate::utils::{match_type, walk_ptrs_ty, paths, span_help_and_lint};
+use if_chain::if_chain;
+use rustc::hir::*;
+use rustc::lint::{LateContext, LateLintPass, LintArray, LintPass};
+use syntax::ast::{LitKind};
+use rustc::{declare_tool_lint, lint_array};
+use std::path::Path;
+
+declare_clippy_lint! {
+    /// **What it does:*** Checks for `push` calls on `PathBuf` that can cause overwrites.
+    ///
+    /// **Why is this bad?** Calling `push` with a root path at the start can overwrite the
+    /// previous defined path.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```ignore
+    /// let mut x = PathBuf::from("/foo");
+    /// x.push("/bar");
+    /// println!("x: {}", x.display()); // "/bar
+    /// ```
+    /// Could be written:
+    ///
+    /// ```ignore
+    /// let mut x = PathBuf::from("/foo");
+    /// x.push("bar");
+    /// println!("x: {}", x.display()); // "/foo/bar
+    /// ```
+    pub PATH_BUF_PUSH_OVERWRITE,
+    style,
+    "calling `push` with file system root on `PathBuf` can overwrite it"
+}
+
+#[derive(Copy, Clone)]
+pub struct Pass;
+
+impl LintPass for Pass {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(PATH_BUF_PUSH_OVERWRITE)
+    }
+
+    fn name(&self) -> &'static str {
+        "PathBufPushOverwrite"
+    }
+}
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for Pass {
+    fn check_expr(&mut self, cx: &LateContext<'a, 'tcx>, expr: &'tcx Expr) {
+        if_chain! {
+            if let ExprKind::MethodCall(ref path, _, ref args) = expr.node;
+            if path.ident.name == "push";
+            if match_type(cx, walk_ptrs_ty(cx.tables.expr_ty(&args[0])), &paths::PATH_BUF);
+            if let Some(get_index_arg) = args.get(1);
+            if let ExprKind::Lit(ref lit) = get_index_arg.node;
+            if let LitKind::Str(ref r, _) = lit.node;
+            then {
+                if Path::new(&r.as_str()).has_root() {
```

Comment:
Would [`is_absolute`](https://doc.rust-lang.org/1.27.2/std/path/struct.PathBuf.html#method.is_absolute) be better than `has_root` here? The docs for [`push`](https://doc.rust-lang.org/1.27.2/std/path/struct.PathBuf.html#method.push) seem to suggest the path will be replaced if the other path is absolute.
Also, could this `if` be part of the `if_chain!` above?

Reply:
That makes sense! Seems like `is_absolute` is better. As It's only equivalent to `is_root` on Unix. I'll make the changes.

Reply:
Copying from below: https://github.com/rust-lang/rust-clippy/pull/3954#discussion_r275237252
> I'm not sure about the is_absolute vs has_root thing. ::push("\windows") on windows will result in c:\windows, while it doesn't matter what the previous path was (except the prefix). But is_absolute does not cover this case.